PETSc partitioner

fdkong opened the pull request:
We are supporting all PETSc external partitioners, `parmetis`, `chaco`, `party` and `ptscotch`.

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

fdkong committed:
Added MatPartitioning interface

Refs #11628

fdkong committed:
Add flags in TestHarness for chaco, parmetis, ptscotch and party

Refs #11628

fdkong committed:
Added tests for PETSc partitioner chaco, parmetis, ptscotch, and party

Refs #11628

fdkong committed:
Add support hierarchical partitioning

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/204060/) on 87686c7 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11741/site)

This comment will be updated on new commits.

fdkong committed:
Restrict chaco to 32-bit integers

Refs #11628

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit ceb1e02bee97b3758bfcdc37cb87bc8954a89030
+Subproject commit 0bc2791720d43ea8fa8dc3fd5193aacbc1847669
```
**Caution!** This contains a submodule update

fdkong committed:
Fixed TestHarness for chaco, party, ptscotch, and parmetis

Refs #11628

fdkong committed:
Explicitly require PETSc-3.9.3 or higher for party and chaco

because I added patch in PETSc to allow the serial partitioners
use the distributed graph

Refs  #11628

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/204487/) on 6df4b6a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11741/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11741/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5a2f370d09c9355f4349998d80817e81cf0234b4`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/204498/) on c874988 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11741/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11741/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5a2f370d09c9355f4349998d80817e81cf0234b4`

fdkong committed:
Added docs for PETSc partitioner interfaces

Refs  #11628

fdkong committed:
PetscMatPartitioner: Added design requirement in 'tests'

Refs  #11628

fdkong commented:
@friedmud  Could you consider merging this PR? Then your changes can be based on this. I will add the weighting interface based on your changes.

Please let me know what do you think.

fdkong committed:
PetscMatPartitioning: add interface to specify element and side weights

Refs #11628

fdkong commented:
Waiting for Libmesh update

fdkong committed:
Add actual weighting paritioner tests

Refs #11628

permcody commented on test/tests/partitioners/petsc_partitioner/tests:
```diff
@@ -0,0 +1,137 @@
+[Tests]
+  [./ptscotch]
+    requirement = 'MOOSE need more partitioning options'
```
These are not good requirement statements. I would suggest saying something like this instead.
"MOOSE shall support the <partitioner> mesh partitioner"

Requirements should be specific, testable, unique, etc:
https://qracorp.com/write-clear-requirements-document/

moosebuild commented:
All [jobs](https://civet.inl.gov/event/31441/) on 3e19cb6 : invalidated by @fdkong

Libmesh updates are done

fdkong committed:
Added documentation for PartitionerWeight.C

Refs #11628

fdkong committed:
Add more docs for PetscMatPartitioner

Refs #11628

fdkong committed:
Added pictures for PetscMatPartitioner.md

Refs #11628

friedmud commented on framework/doc/content/documentation/systems/Mesh/Partitioner/PartitionerWeight.md:
```diff
@@ -0,0 +1,7 @@
+# PartitionerWeight
+
+It is a test `partitioner` to demonstrate how to assign the vertex weights and the edge weights to the graph.
```
Let's remove `vertex` from the documentation... that means "nodes" to many people.  Let's just talk about "element weights" and "side weights" for now.

friedmud commented on framework/doc/content/documentation/systems/Mesh/Partitioner/PartitionerWeight.md:
```diff
@@ -0,0 +1,7 @@
+# PartitionerWeight
+
+It is a test `partitioner` to demonstrate how to assign the vertex weights and the edge weights to the graph.
+The vertex weights (corresponding to word load) could be assigned by overriding `computeElementWeight`. The edge
```
`word` -> `work`

friedmud commented on test/include/partitioner/PartitionerWeight.h:
```diff
@@ -0,0 +1,38 @@
+#include "MooseEnum.h"
+#include "PetscMatPartitioner.h"
+
+class PartitionerWeight;
```
Can you please name this `PartitionerWeightTest`... just so that it's clear that it's a "test" object?

friedmud reviewed (changes requested):
Ok - here's a big one: can we rename `PetscMatPartitioner` to just `PetscPartitioner`?  The fact that it's through the "mat" interface in PETSc doesn't mean anything to users of MOOSE... all we care is that we are using PETSc's interface to the partitioners.

fdkong committed:
Change PetscMatPartitioner to PetscExternalPartitioner

Refs #11628

fdkong committed:
Renamed PartitionerWeight to PartitionerWeightTest

fdkong committed:
Changed API computeSideWeight

Refs #11628

fdkong commented:
Any more suggestions @friedmud
